azkaban-common/src/main/java/azkaban/utils/OsCpuUtil.java

```diff
@@ -0,0 +1,195 @@
+/*
+ * Copyright 2019 LinkedIn Corp.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not
+ * use this file except in compliance with the License. You may obtain a copy of
+ * the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations under
+ * the License.
+ */
+
+package azkaban.utils;
+
+import java.io.BufferedReader;
+import java.io.IOException;
+import java.nio.charset.StandardCharsets;
+import java.nio.file.Files;
+import java.nio.file.Paths;
+import java.util.ArrayDeque;
+import java.util.Deque;
+import javax.inject.Inject;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+
+/**
+ * Utility class for getting CPU usage (in percentage)
+ *
+ * CPU information is obtained from /proc/stat, so only Linux systems will support this class
+ *
+ * Assumes frequent calls at regular intervals to {@link #getCpuLoad() getCpuLoad}. The length of
+ * time over which cpu load is calculated can be adjusted with parameter
+ * {@code numCpuStatsToCollect} in the class constructor and how often {@link #getCpuLoad()
+ * getCpuLoad} is called.
+ * Example: if {@link #getCpuLoad() getCpuLoad} is called every second and {@code
+ * numCpuStatsToCollect} is set to 60 then each call to {@link #getCpuLoad() getCpuLoad} returns
+ * the cpu load over the last minute.
+ */
+public class OsCpuUtil {
+
+  private static final Logger logger = LoggerFactory.getLogger(OsCpuUtil.class);
+
+  private static final String CPU_STAT_FILE = "/proc/stat";
+
+  private final Deque<CpuStats> collectedCpuStats;
+
+  @Inject
+  public OsCpuUtil(int numCpuStatsToCollect) {
+    if (numCpuStatsToCollect <= 0) {
+      numCpuStatsToCollect = 1;
+    }
+    this.collectedCpuStats = new ArrayDeque<>(numCpuStatsToCollect);
+
+    final CpuStats cpuStats = getCpuStats();
+    if (cpuStats != null) {
+      for (int i = 0; i < numCpuStatsToCollect; i++) {
+        this.collectedCpuStats.push(cpuStats);
+      }
+    }
+  }
+
+  /**
+   * Collects a new cpu stat data point and calculates cpu load with it and the oldest one
+   * collected which is then deleted.
+   *
+   * @return percentage of CPU usage. -1 if there are no cpu stats.
+   */
+  public double getCpuLoad() {
```

Comment:
Is calculating the newest and oldest data points the right way to get CPU load? Let's discuss more about it.

Reply:
The data points are the pairs (system uptime and cpu idle time since last boot). By taking the difference between a data point from now and another from 5 minutes ago, for example, we are calculating what happened during that time in terms of cpu usage, basically we are saying over the last 5 minutes how much of this time was the cpu idle as a percentage, the remaining percentage would then indicate the cpu usage.